source/code/plugins/in_npmd.rb

```diff
@@ -0,0 +1,676 @@
+                Logger::logInfo "Restarting NPMD"
+                start_npmd_async_retry_thread()
+            end
+        end
+
+        def start_npmd
+            unless is_npmd_seen_in_ps()
+                @reader, @writer = IO.pipe
+                @npmdIntendedStop = false
+                @npmdProcessId = Process.spawn(Loc_npmd_binary, :err=>@writer)
+                _t = Thread.new(&method(:handle_exit))
+            else
+                Logger::logInfo "Npmd already seen in PS"
+            end
+        end
+
+        def stop_npmd
+            if defined?@npmdProcessId and is_npmd_seen_in_ps()
+                @stop_sync.synchronize do
+                    @npmdIntendedStop = true
+                    Process.kill(Stop_signal, @npmdProcessId.to_i)
+                end
+            else
+                Logger::logInfo "NPMD agent found already stopped"
+            end
+        end
+
+        def send_config
+            begin # checking for File call errors with this config file
+                unless File.exist?(Loc_config_file)
+                    Logger::logWarn "No file #{Loc_config_file} found at location"
+                else
+                    if defined?@npmdClientSock and !@npmdClientSock.nil?
+
+                        # Read the UI configuration from file location
+                        _uiXml = File.read(Loc_config_file)
+                        if _uiXml.nil? or _uiXml == ""
+                            Logger::logWarn "File read at #{Loc_config_file} got nil or empty string"
+                            return
+                        end
+
+                        # Transform the UI XML configuration to agent configuration
+                        _agentConfig = NPMDConfig::GetAgentConfigFromUIConfig(_uiXml)
+                        if _agentConfig.nil? or _agentConfig == ""
+                            Logger::logWarn "Agent configuration transformation returned empty"
+                            return
+                        end
+
+                        @npmdClientSock.puts _agentConfig
+                        @npmdClientSock.flush
+
+                    else
+                        Logger::logWarn "NPMD client socket not connected yet!"
+                    end
+                end
+            rescue Exception => e
+                log_error "Exception when sending config: #{e.class} #{e.message}", Logger::resc
+            rescue StandardError => e
+                log_error "Error when sending config: #{e}", Logger::resc
+            end
+        end
+
+        def server_run
+            unless defined?@server_obj and !@server_obj.nil?
+                Logger::logInfo "Server obj was not created properly, Exiting"
+            else
+                @fqdn = OMS::Common.get_fully_qualified_domain_name()
+                @agentId = OMS::Configuration.agent_id()
+                # Wait for agentId to be set to valid value
+                wait_for_agent_id()
```

Comment:
NPMD agent needs a GUID to handle the aggregation of data. The original idea was to re-use the one generated from OMS::Configuration, but based on the dependency issue highlighted, it is easy for us to maintain our own GUID. We will make necessary change and avoid plugin dependency.